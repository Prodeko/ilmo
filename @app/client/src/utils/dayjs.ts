import { Locale } from "antd/lib/locale"
import enUS from "antd/lib/locale/en_US"
import fiFI from "antd/lib/locale/fi_FI"
import svSE from "antd/lib/locale/sv_SE"
import dayjs from "dayjs"
import timezone from "dayjs/plugin/timezone"
import utc from "dayjs/plugin/utc"

import "dayjs/locale/fi"
import "dayjs/locale/en"
import "dayjs/locale/sv-fi"

// Required for dayjs timezone features
// https://day.js.org/docs/en/plugin/timezone
dayjs.extend(utc)
dayjs.extend(timezone)

const locales = {
  fi: ["fi", fiFI],
  en: ["en", enUS],
  se: ["se", svSE],
}

export function setLocale(locale: string): Locale {
  const [localeString, antd] = locale ? locales[locale] : [locales["fi"]]
  dayjs.locale(localeString)
  return antd
}
