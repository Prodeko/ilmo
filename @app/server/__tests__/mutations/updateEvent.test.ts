import dayjs from "dayjs";

import {
  asRoot,
  createEventDataAndLogin,
  deleteTestData,
  runGraphQLQuery,
  sanitize,
  setup,
  teardown,
} from "../helpers";

beforeEach(deleteTestData);
beforeAll(setup);
afterAll(teardown);

describe("UpdateEvent", () => {
  it("can update an existing event", async () => {
    const {
      events,
      organization,
      eventCategory,
      session,
    } = await createEventDataAndLogin();
    const event = events[0];

    const day = dayjs("2021-02-20T12:00:00+02:00");

    await runGraphQLQuery(
      `mutation UpdateEvent(
        $eventId: UUID!
        $name: JSON!
        $description: JSON!
        $ownerOrganizationId: UUID!
        $categoryId: UUID!
        $isHighlighted: Boolean
        $isDraft: Boolean
        $eventStartTime: Datetime!
        $eventEndTime: Datetime!
        $registrationStartTime: Datetime!
        $registrationEndTime: Datetime!
      ) {
        updateEvent(
          input: {
            id: $eventId
            patch: {
              name: $name
              description: $description
              ownerOrganizationId: $ownerOrganizationId
              categoryId: $categoryId
              isHighlighted: $isHighlighted
              isDraft: $isDraft
              eventStartTime: $eventStartTime
              eventEndTime: $eventEndTime
              registrationStartTime: $registrationStartTime
              registrationEndTime: $registrationEndTime
            }
          }
        ) {
          event {
            id
            name
            description
            ownerOrganizationId
            categoryId
            isHighlighted
            isDraft
            eventStartTime
            eventEndTime
            registrationStartTime
            registrationEndTime
            createdBy
            updatedBy
            createdAt
            updatedAt
          }
        }
      }`,

      // GraphQL variables:
      {
        eventId: event.id,
        slug: event.slug,
        name: { fi: "P채ivitetty testitapahtuma", en: "Updated test event" },
        description: {
          fi: "P채ivitetty testikuvaus",
          en: "Updated test description",
        },
        ownerOrganizationId: organization.id,
        categoryId: eventCategory.id,
        isHighlighted: true,
        isDraft: false,
        eventStartTime: day.add(2, "hour").toISOString(),
        eventEndTime: day.add(3, "hour").toISOString(),
        registrationStartTime: day.toISOString(),
        registrationEndTime: day.add(1, "hour").toISOString(),
      },

      // Additional props to add to `req` (e.g. `user: {session_id: '...'}`)
      {
        user: { session_id: session.uuid },
      },

      // This function runs all your test assertions:
      async (json, { pgClient }) => {
        expect(json.errors).toBeFalsy();
        expect(json.data).toBeTruthy();

        const updatedEvent = json.data!.updateEvent.event;

        expect(updatedEvent).toBeTruthy();
        expect(updatedEvent.ownerOrganizationId).toEqual(organization.id);

        expect(sanitize(updatedEvent)).toMatchInlineSnapshot(`
          Object {
            "categoryId": "[id-3]",
            "createdAt": "[timestamp-5]",
            "createdBy": "[id-4]",
            "description": Object {
              "en": "Updated test description",
              "fi": "P채ivitetty testikuvaus",
            },
            "eventEndTime": "[timestamp-2]",
            "eventStartTime": "[timestamp-1]",
            "id": "[id-1]",
            "isDraft": false,
            "isHighlighted": true,
            "name": Object {
              "en": "Updated test event",
              "fi": "P채ivitetty testitapahtuma",
            },
            "ownerOrganizationId": "[id-2]",
            "registrationEndTime": "[timestamp-4]",
            "registrationStartTime": "[timestamp-3]",
            "updatedAt": "[timestamp-6]",
            "updatedBy": "[id-4]",
          }
        `);

        const { rows } = await asRoot(pgClient, () =>
          pgClient.query(`SELECT * FROM app_public.events WHERE id = $1`, [
            updatedEvent.id,
          ])
        );

        if (rows.length !== 1) {
          throw new Error("Event not found!");
        }
        expect(rows[0].id).toEqual(updatedEvent.id);
      }
    );
  });

  it("can't update an event while logged out (RLS policy)", async () => {
    const { events } = await createEventDataAndLogin({
      registrationOptions: { create: false },
    });
    const eventId = events[0].id;

    await runGraphQLQuery(
      `mutation UpdateEvent(
        $eventId: UUID!
        $description: JSON!
      ) {
        updateEvent(
          input: {
            id: $eventId
            patch: {
              description: $description
            }
          }
        ) {
          event {
            id
          }
        }
      }`,

      // GraphQL variables:
      {
        eventId,
        description: "Test",
      },

      // Additional props to add to `req` (e.g. `user: {session_id: '...'}`)
      {},

      // This function runs all your test assertions:
      async (json) => {
        expect(json.errors).toBeTruthy();

        const message = json.errors![0].message;
        expect(message).toEqual(
          "No values were updated in collection 'events' because no values you can update were found matching these criteria."
        );
      }
    );
  });
});
