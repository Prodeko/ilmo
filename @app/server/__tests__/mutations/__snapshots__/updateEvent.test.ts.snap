// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpdateEvent can update an existing event 2`] = `
Object {
  "data": Object {
    "updateEvent": Object {
      "event": Object {
        "categoryId": "[id-3]",
        "createdAt": "[timestamp-5]",
        "createdBy": "[id-4]",
        "description": Object {
          "en": "Updated test description",
          "fi": "Päivitetty testikuvaus",
        },
        "eventEndTime": "[timestamp-2]",
        "eventStartTime": "[timestamp-1]",
        "id": "[id-1]",
        "isDraft": false,
        "isHighlighted": true,
        "name": Object {
          "en": "Updated test event",
          "fi": "Päivitetty testitapahtuma",
        },
        "ownerOrganizationId": "[id-2]",
        "registrationEndTime": "[timestamp-4]",
        "registrationStartTime": "[timestamp-3]",
        "updatedAt": "[timestamp-6]",
        "updatedBy": "[id-4]",
      },
    },
  },
}
`;

exports[`UpdateEvent can't update an event while logged out (RLS policy) 1`] = `
Object {
  "data": Object {
    "updateEvent": null,
  },
  "errors": Array [
    Object {
      "extensions": Object {
        "exception": Object {},
      },
      "locations": Array [
        Object {
          "column": 9,
          "line": 5,
        },
      ],
      "message": "No values were updated in collection 'events' because no values you can update were found matching these criteria.",
      "path": Array [
        "updateEvent",
      ],
    },
  ],
}
`;
